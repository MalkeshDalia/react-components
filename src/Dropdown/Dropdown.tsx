import * as React from "react";
import Select, { components } from "react-select";
import * as SelectNS from "react-select/lib/Select";
import * as cx from "classnames";

export namespace Dropdown {
  export type Props<OptionType> = SelectNS.Props<OptionType> & { size: "medium" | "small", flat: boolean };
}

const defaultComponents: Dropdown.Props<any>["components"] = {
  Placeholder: ({ children }) => (
    <div className="dropdown__placeholder">{children}</div>
  ),
  IndicatorSeparator: () => null,
  Option: props => (
    <components.Option
      {...props}
      className={cx({
        "is-disabled": props.isDisabled,
        "is-focused": props.isFocused,
        "is-selected": props.isSelected
      })}
    />
  ),
  DropdownIndicator: props => (
    <span className="dropdown__select-arrow-zone">
      <span
        className={
          props.selectProps.menuIsOpen
            ? "dropdown__select-arrow-up"
            : "dropdown__select-arrow-down"
        }
      />
    </span>
  ),
  MultiValueLabel: ({ children }) => (
    <div className="dropdown__multi-value-label">{children}</div>
  ),
  MultiValueRemove: props => (
    <components.MultiValueRemove
      innerProps={{
        ...props.innerProps,
        className: "dropdown__multi-value-remove"
      }}
    >
      ×
    </components.MultiValueRemove>
  ),
  ClearIndicator: props => (
    <span onClick={props.clearValue} className="dropdown__clear-indicator">
      ×
    </span>
  )
};

export class Dropdown<OptionType> extends React.Component<
  Dropdown.Props<OptionType>
> {
  static defaultProps: Dropdown.Props<any> = {
    delimiter: ",",
    size: "medium",
    isDisabled: false,
    isSearchable: false,
    isClearable: false,
    isMulti: false,
    flat: false,
    blurInputOnSelect: true,
    menuPlacement: "bottom",
    components: {}
  };

  getCustomClassNames() {
    const {
      size,
      flat,
      isClearable,
      isMulti,
      isDisabled
    } = this.props;
    return cx({
      "is-medium": size === "medium",
      "is-small": size === "small",
      "is-flat": flat,
      "is-multi": isMulti,
      "is-clearable": isClearable,
      "is-disabled": isDisabled
    });
  }

  private select: Select<OptionType> | null;

  focus = () => {
    this.select && this.select.focus();
  };

  render() {
    const {
      props: { className, components: customComponents, ...props }
    } = this;

    return (
      <Select
        {...props}
        classNamePrefix="dropdown"
        components={{
          ...defaultComponents,
          ...customComponents
        }}
        className={cx("dropdown", className, this.getCustomClassNames())}
        ref={select => {
          this.select = select;
        }}
      />
    );
  }
}
