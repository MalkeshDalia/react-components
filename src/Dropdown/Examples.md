### Examples

```js
initialState = {}

const onChange = (value) => setState({ value });

const options = [
  { value: 'apple', label: 'Apple' },
  { value: 'avocado', label: 'Avocado' },
  { value: 'orange', label: 'Orange' },
  { value: 'lemon', label: 'Lemon' },
  { value: 'mandarin', label: 'Mandarin' }
];

<Dropdown
  value={state.value}
  onChange={onChange}
  placeholder='Select some fruit'
  options={options}
/>
```


#### Dropdowns sizes
Dropdown comes in two different sizes: medium and small
```js
initialState = {}

const onChange = (stateVar) => (value) => setState({ [stateVar]: value });

const { value, value2 } = state;

const options = [
  { value: 'apple', label: 'Apple' },
  { value: 'avocado', label: 'Avocado' },
  { value: 'orange', label: 'Orange' },
  { value: 'lemon', label: 'Lemon' },
  { value: 'mandarin', label: 'Mandarin' }
];
const options2 = [
  { value: 'cabbage', label: 'Cabbage' },
  { value: 'carrot', label: 'Carrot' },
  { value: 'peas', label: 'Peas' },
  { value: 'tomato', label: 'Tomato' },
  { value: 'cucumber', label: 'Cucumber' }
];

<FlexView vAlignContent='center' className='dropdown-list'>
  <Dropdown
    value={state.value}
    onChange={onChange('value')}
    placeholder='Select some fruit'
    options={options}
  />
  <Dropdown
    value={value2}
    onChange={onChange('value2')}
    placeholder='Select some vegetables'
    options={options2}
    size='small'
  />
</FlexView>
```


#### Menu position
When needed (e.g. when there isn't enough space under the dropdown to render the menu), it's possible to force the menu to open on top
```js
intialState = {};

const onChange = (value) => setState({ value });

const options = [
  { value: 'apple', label: 'Apple' },
  { value: 'avocado', label: 'Avocado' },
  { value: 'orange', label: 'Orange' },
  { value: 'lemon', label: 'Lemon' },
  { value: 'mandarin', label: 'Mandarin' }
];

<Dropdown
  value={state.value}
  onChange={onChange}
  placeholder='Select some fruit'
  options={options}
  menuPosition='top'
/>
```


#### Dropdowns options
* When the `searchable` prop is used, it's possible to search the desired value by writing into the dropdown.\n* Use `Clearable` if it should be possible to reset the selected value clicking on the `delete` icon that will appear once a value is selected.
```js
intialState = {};

const onChange = (value) => setState({ value });

const options = [
  { value: 'apple', label: 'Apple' },
  { value: 'avocado', label: 'Avocado' },
  { value: 'orange', label: 'Orange' },
  { value: 'lemon', label: 'Lemon' },
  { value: 'mandarin', label: 'Mandarin' },
  { value: 'lime', label: 'Lime' },
  { value: 'peach', label: 'Peach' },
  { value: 'apricot', label: 'Apricot' },
  { value: 'pineapple', label: 'Pineapple' },
  { value: 'banana', label: 'Banana' }
];

<Dropdown
  className='custom'
  value={state.value}
  onChange={onChange}
  searchable
  clearable
  backspaceRemoves
  placeholder='Select some fruit (try to type "Banana")'
  options={options}
/>
```


#### Multiselect
Dropdown could also allow the selection of multiple values
```js
intialState = {};

const onChange = (value) => setState({ value });

const options = [
  { value: 'apple', label: 'Apple' },
  { value: 'avocado', label: 'Avocado' },
  { value: 'orange', label: 'Orange' },
  { value: 'lemon', label: 'Lemon' },
  { value: 'mandarin', label: 'Mandarin' }
];

<Dropdown
  className='custom'
  value={state.value}
  onChange={onChange}
  multi
  clearable
  placeholder='Select some fruit(s)'
  options={options}
/>
```


#### Groups
Dropdown could also allow the grouping
```js
initialState = {}

const onChange = (value) => setState({ value });

const options = [
  { value: 'cucumber', label: 'Cucumber', category: 'Vegetable'},
  { value: 'apple', label: 'Apple', category: 'Fruit'},
  { value: 'orange', label: 'Orange', category: 'Fruit' },
  { value: 'mandarin', label: 'Mandarin', category: 'Fruit' },
  { value: 'onion', label: 'Onion', category: 'Vegetable' },
  { value: 'peach', label: 'Peach', category: 'Fruit' },
  { value: 'apricot', label: 'Apricot', category: 'Fruit' },
  { value: 'tomato', label: 'Tomato', category: 'Vegetable' },
  { value: 'pineapple', label: 'Pineapple', category: 'Fruit' },
  { value: 'banana', label: 'Banana', category: 'Fruit'},
  { value: 'meat', label: 'Meat' }
];

<Dropdown
  className='custom'
  value={state.value}
  onChange={onChange}
  searchable
  clearable
  backspaceRemoves
  placeholder='Select some fruit (try to type "Banana")'
  options={options}
  groupByKey='category'
/>
```
