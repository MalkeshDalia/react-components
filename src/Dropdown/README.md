# Dropdown
Il dropdown permette all’utente di selezionare uno o più valori all’interno di una lista. 

È consigliabile utilizzare il dropdown quando l’utente deve scegliere tra una lista numerosa di elementi. Nel caso di liste al di sotto di 5 elementi, è preferibile utilizzare radio buttons o checkboxes.

![defaultdropdowncover](https://user-images.githubusercontent.com/10867086/35731278-40952d0a-0815-11e8-8e2f-556387943587.jpg)

## Stati
|Name|Type|Description|
|----|----|-----------|
|![defaultdropdown](https://user-images.githubusercontent.com/10867086/35731368-97af8478-0815-11e8-9389-2f328cfbf89f.jpg)|**Normal**|Nessuna interazione da parte dell’utente, il Dropdown  è abilitato e può essere cliccato.|
|![dropdownhover](https://user-images.githubusercontent.com/10867086/35731426-c8b22c2e-0815-11e8-8dd5-9c93f7c5bf4f.jpg)|**Hover**|L’utente è in hover sul Dropdown, che è attivo e può essere cliccato.|
|![opendropdown](https://user-images.githubusercontent.com/10867086/35731466-eb3aaba4-0815-11e8-9018-3e9bc4842746.jpg)|**Open**|L’utente ha cliccato sul dropdown, che mostra una listra di elementi selezionabili. A seconda dello spazio disponibile, il pannello può aprirsi sopra o sotto il Dropdown|
|![hasvaluedropdown](https://user-images.githubusercontent.com/10867086/35731797-5835f0b4-0817-11e8-8798-f7c45f07e628.jpg)|**Has Value**|L’utente ha selezionato uno o più valori dalla lista, che vengono visualizzati all’interno del Dropdown|
|![disableddropdown](https://user-images.githubusercontent.com/10867086/35731836-805ba91c-0817-11e8-89e1-7443bf088206.jpg)|**Disabled**|Indipendentemente dall’azione dell’utente, il Dropdown è disabilitato e non può essere cliccato.|

## Varianti
|Name|Type|Description|
|----|----|-----------|
|![clearabledropdown](https://user-images.githubusercontent.com/10867086/35732083-701b0b6e-0818-11e8-9625-7c1512284cc5.jpg)|**Clearable**|L’ utente può rimuovere il valore selezionato grazie ad un’ icona posta all’interno del dropdown.|
|![searchabledropdown](https://user-images.githubusercontent.com/10867086/35732084-73f5663a-0818-11e8-8a7b-06cba30bb06e.jpg)|**Searchable**|L’ utente può cercare un valore digitando del testo all’interno del dropdown. Per fare in modo che l’utente capisca che il dropdown è searchable nello stato di hover utilizzare il cursore “text”|
|![multiselectdropdown](https://user-images.githubusercontent.com/10867086/35732088-772ed2a0-0818-11e8-885a-9397d13e484a.jpg)|**Multiselect**|L’ utente può selezionare più valori, che verranno visualizzati all’interno del dropdown. Una volta selezionati i valori potranno essere rimossi ad uno ad uno, oppure tutti insieme (I dropdown multiselect sono solitamente anche clearable)|
|![groupsdropdown](https://user-images.githubusercontent.com/10867086/35732097-7ea6a51c-0818-11e8-99cd-f9119eaba44e.jpg)|**Groups**|Per facilitare l’utente nella ricerca, gli elementi della lista possono essere divisi in gruppi separati attraverso un titolo.|

## Dimensioni
Il Dropdown, a seconda delle esigenze può essere di due dimensioni:

![dropdownsize](https://user-images.githubusercontent.com/10867086/35732225-f0ddff18-0818-11e8-8ec8-e981248e0fba.jpg)
