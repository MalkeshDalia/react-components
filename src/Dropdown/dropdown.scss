@import "~sass-flex-mixins/src/sass-flex-mixins.scss";
@import "../utils/theme/variables.scss";
$background: $brc-gradientGrey !default;
$background-hover: $brc-gradientDarkGrey !default;
$background-open: $brc-gradientDarkGrey !default;
$border-color: $brc-silver !default;
$border-color-hover: $border-color !default;
$border-color-open: $brc-waterBlue !default;
$border-color-focus: $brc-waterBlue !default;
$border-radius: 2px !default;
$text-color: $brc-coolGrey !default;
$text-color-open: $brc-coolGrey !default;
$text-color-has-value: $brc-darkGreyBlue !default;
$text-color-hover: $text-color-has-value !default;
$text-color-writing: $brc-coolGrey !default;
$placeholder-color: $brc-coolGrey !default;
$placeholder-color-hover: $brc-coolGrey !default;
$placeholder-color-open: $brc-coolGrey !default;
$arrow-color: $text-color !default;
$arrow-color-hover: $text-color !default;
$arrow-color-open: $text-color-open !default;
$x-color: $text-color !default;
$x-color-hover: $brc-red !default;
$menu-background: $brc-white !default;
$menu-border-color: $brc-silver !default;
$option-group-color: $brc-darkGreyBlue !default;
$option-group-background: $brc-paleGrey !default;
$option-color: rgba($brc-darkGreyBlue, 0.5) !default;
$option-color-hover: $brc-darkGreyBlue !default;
$option-color-selected: $option-color !default;
$option-color-disabled: rgba($option-color, 0.2) !default;
$option-background-hover: rgba($brc-azure, 0.15) !default;
$option-background-selected: $menu-background !default;
$option-background-disabled: $menu-background !default;
$multiple-value-color: $brc-azure !default;
$multiple-value-border-color: $multiple-value-color !default;
$multiple-value-background-color: $brc-duckBlue !default;
$multiple-value-icon-color: $multiple-value-color !default;
$multiple-value-icon-background-color: $multiple-value-background-color !default;
$multiple-value-icon-border-color: $multiple-value-icon-color !default;
$multiple-value-icon-color-hover: $brc-red !default;
$multiple-value-icon-background-color-hover: darken(
  $multiple-value-icon-background-color,
  10%
) !default;
$multiple-value-icon-border-color-hover: $multiple-value-icon-border-color !default;
$disabled-opacity: 0.6 !default;
$height: 36px !default;
$height-small: 30px !default;
$padding-left-small: 10px !default;
$padding-left: 10px !default;

// NORMALIZE
@mixin normalize {
  box-sizing: content-box;
  margin: 0;
  padding: 0;

  @include display_flex();
  @include flex_flex-direction(row);
  @include flex_align-items(center);
}

@mixin height($height) {
  height: $height;
  line-height: $height;
}

.buildo-dropdown {
  .Select-control {
    min-height: 0;
    box-sizing: border-box;
    @include normalize();
    @include height($height);
  }
  .Multi-value,
  .Single-value,
  .Select-placeholder,
  .Select-input,
  .Select-control > input,
  .Select-placeholder,
  .Select-option,
  .Select-arrow-zone,
  .Clear-indicator {
    @include normalize();
    @include height($height - 2);
  }

  &.is-small {
    .Select-control {
      @include height($height-small);
    }
    .Multi-value,
    .Single-value,
    .Select-placeholder,
    .Select-input,
    .Select-control > input,
    .Select-placeholder,
    .Select-option,
    .Select-arrow-zone,
    .Clear-indicator {
      @include height($height-small - 2);
    }
  }
}

.buildo-dropdown {
  min-width: 150px;
  background: $background;
  color: $text-color;
  font-size: 15px;
  font-weight: $brc-semibold;
  height: 100%;

  &:hover {
    border-color: $border-color-hover;
  }

  &:not(.is-disabled) {
    .Select-control.is-focused {
      border-color: $border-color-focus;
    }
  }

  &.is-multi {
    height: auto;
    .Select-control {
      height: auto;
    }
  }

  .Select-control {
    border: solid 1px $border-color;
    border-radius: $border-radius;
    background: transparent;
    cursor: pointer;
    box-shadow: none !important;
    flex-grow: 1;

    .Select-multi-value-wrapper {
      @include flex_flex-wrap(wrap);
      @include flex_align-items(center);
    }

    .Select-input,
    .Select-placeholder {
      background: transparent;
    }

    .Select-arrow-zone,
    .Clear-indicator {
      @include flex_flex(0 0 auto);
      @include flex_justify-content(center);
    }

    .Select-arrow-zone {
      height: 100%;
      width: 30px;
    }

    .Clear-indicator {
      width: 15px;
    }

    .Select-arrow,
    .Clear-indicator {
      cursor: pointer;
      font-size: 20px;
    }

    .Single-value,
    .Multi-value {
      font-size: 0.9em;
    }

    .Select-arrow {
      border-color: $arrow-color transparent transparent;
      border-width: 5px 5px 0;
    }
  }

  .Select-placeholder {
    font-weight: $brc-semibold;
    color: $placeholder-color;
  }

  .Select-menu-outer {
    position: absolute;
    left: 0;
    z-index: 1000;
    margin-top: 10px;
    border: solid 1px $menu-border-color;
    border-radius: 4px;
    min-width: 185px;
    background-color: $menu-background;
    box-shadow: rgba(34, 36, 38, 0.12) 0 2px 4px 0;

    .Select-option-group {
      font-size: 15px;
      font-weight: $brc-semibold;
      color: $option-group-color;
      background: $option-group-background;
    }

    .Select-option {
      padding: 0 10px;
      cursor: pointer;
      font-size: 15px;
      font-weight: 400;
      color: $option-color;
      background: $menu-background;

      &.is-selected {
        background: $option-background-selected;
        color: $option-color-selected;
      }

      &.is-disabled {
        color: $option-color-disabled;
        background: $option-background-disabled;
      }

      &:not(.is-disabled):hover,
      &.is-focused {
        background: $option-background-hover;
        color: $option-color-hover;
      }
    }
  }

  &.menu-position-top {
    .Select-menu-outer {
      top: auto;
      bottom: 100%;
      margin-bottom: 10px;
    }
  }

  .Multi-value {
    margin-right: 5px;
    margin-top: 2px;
    margin-bottom: 2px;
    border: 1px solid $multiple-value-border-color;
    flex-direction: row-reverse;

    @include height($height - 14px);

    .Multi-value-label {
      @include height($height - 14px);

      background: $multiple-value-background-color;
      color: $multiple-value-color;
      padding: 0 5px;
    }

    .Multi-value-remove {
      padding: 0 4px;
      border-color: $multiple-value-icon-border-color;
      background: $multiple-value-icon-background-color;
      color: $multiple-value-icon-color;
      border-right: 1px solid;
      display: inline-block;
      vertical-align: middle;

      &:hover {
        border-color: $multiple-value-icon-border-color-hover;
        background: $multiple-value-icon-background-color-hover;
        color: $multiple-value-icon-color-hover;
      }
    }
  }

  &.is-small {
    .Select-input,
    .Select-placeholder {
      @include height($height-small - 12px);
    }

    .Multi-value {
      @include height($height-small - 14px);

      .Multi-value-label {
        @include height($height-small - 14px);
      }
    }
  }

  &:not(.is-disabled):hover {
    background: $background-hover;

    .Single-value {
      color: $text-color-hover;
    }

    .Select-placeholder {
      color: $placeholder-color-hover;
    }

    .Select-arrow {
      border-color: $arrow-color-hover transparent transparent;
    }
  }

  &:not(.is-disabled) {
    .Select-control.is-open {
      border-color: $border-color-open;
      background: $background-open;

      .Single-value {
        color: $text-color-open;
      }

      .Select-placeholder {
        color: $placeholder-color-open;
      }

      .Select-arrow {
        border-color: $arrow-color-open transparent transparent;
      }
    }
  }

  &.has-value .Select-control {
    .Single-value {
      color: $text-color-has-value;
    }

    .Select-placeholder {
      color: $text-color-has-value;
    }
  }

  .Clear-indicator {
    bottom: 0;
    padding: 0;
    color: $x-color;
    &:hover {
      color: $x-color-hover;
    }
  }

  &.is-searchable {
    .Select-control {
      cursor: text;
    }

    &.is-open {
      .Select-input {
        color: $text-color-writing;
      }
    }
  }

  &.is-disabled {
    cursor: not-allowed;
    opacity: $disabled-opacity;

    .Select-arrow-zone,
    .Select-arrow,
    .Select-control {
      cursor: inherit;
    }
  }

  &.is-flat {
    &:not(:hover),
    &.is-disabled {
      &:not(.is-open) {
        background: transparent;
      }
    }
  }
}
