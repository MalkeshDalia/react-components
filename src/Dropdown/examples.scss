@import "~sass-flex-mixins/src/sass-flex-mixins.scss";
@import "../utils/theme/variables.scss";

.dropdown-list {
  .buildo-dropdown {
    margin-right: 20px;
    min-width: 200px;
  }
}

.buildo-dropdown {
  .Select-arrow-down,
  .Select-arrow-up {
    margin-top: 1px;
    border: 0;
    height: auto;
    width: auto;
    color: $brc-coolGrey;
    font-size: 8px !important;

    @include display_flex();
    @include flex_flex-direction(row);
    @include flex_align-items(center);

    &::before {
      font-family: "icomoon";
    }
  }

  .Select-arrow-up {
    &::before {
      content: "\e910";
    }
  }

  .Select-arrow-down {
    &::before {
      content: "\e90f";
    }
  }
}
