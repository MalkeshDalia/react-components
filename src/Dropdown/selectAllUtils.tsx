import * as React from "react";
import { ContainerProps } from "react-select/lib/components/containers";
import { components } from "react-select";
import { SelectComponentsConfig } from "react-select/lib/components";
import { OptionProps } from "react-select/lib/components/Option";
import { getOptionValue, getOptionLabel } from "react-select/lib/builtins";

export type SelectAllPropsType<OptionType> = {
  option: OptionType;
  isSelected: boolean;
  onSelectedChange: (isSelected: boolean) => void;
};

const ValueContainerWithSelectAll = <OptionType extends {}>(
  selectAll: SelectAllPropsType<OptionType>
) => ({ children, ...props }: ContainerProps<OptionType>) => {
  return (
    <components.ValueContainer {...props}>
      {selectAll.isSelected ? (
        <components.SingleValue
          {...props}
          data={selectAll.option}
          innerProps={{}}
          isDisabled={false}
        >
          {getOptionLabel(selectAll.option)}
        </components.SingleValue>
      ) : (
        children
      )}
    </components.ValueContainer>
  );
};

const Option = <OptionType extends {}>(
  selectAll: SelectAllPropsType<OptionType>
) => ({ children, ...props }: OptionProps<OptionType>) => {
  return (
    <components.Option
      {...props}
      innerProps={{
        ...props.innerProps,
        onClick: e => {
          selectAll.onSelectedChange(
            getOptionValue(props.data) === getOptionValue(selectAll.option)
          );
          props.innerProps.onClick(e);
          if (getOptionValue(props.data) === getOptionValue(selectAll.option)) {
            props.clearValue();
          }
        }
      }}
    >
      {children}
    </components.Option>
  );
};

export const defaultSelectAllComponents = <OptionType extends {}>(
  selectAll: SelectAllPropsType<OptionType>
): SelectComponentsConfig<OptionType> => ({
  ValueContainer: ValueContainerWithSelectAll<OptionType>(selectAll),
  Option: Option(selectAll)
});
