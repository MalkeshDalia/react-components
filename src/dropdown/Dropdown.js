import React from 'react';
import { props, t, skinnable } from '../utils';
import Select from 'react-select';
import find from 'lodash/find';
import omit from 'lodash/omit';
import groupBy from 'lodash/groupBy';
import map from 'lodash/map';
import sortBy from 'lodash/orderBy';
import FlexView from 'react-flexview';
import cx from 'classnames';
import { warn } from '../utils/log';

const isEmptyArray = x => t.Array.is(x) && x.length === 0;

const defaultOptionGroupRenderer = title => title;

export const Props = {
  value: t.maybe(t.union([t.Number, t.String, t.Object, t.list(t.Object)])),
  valueLink: t.maybe(t.struct({
    value: t.maybe(t.union([t.Number, t.String, t.Object, t.list(t.Object)])),
    requestChange: t.Function
  })),
  onChange: t.maybe(t.Function),
  options: t.list(t.Object),
  size: t.enums.of(['medium', 'small']),
  disabled: t.maybe(t.Boolean),
  searchable: t.maybe(t.Boolean),
  clearable: t.maybe(t.Boolean),
  backspaceRemoves: t.maybe(t.Boolean),
  multi: t.maybe(t.Boolean),
  flat: t.maybe(t.Boolean),
  autoBlur: t.maybe(t.Boolean),
  simpleValue: t.maybe(t.Boolean),
  menuPosition: t.enums.of(['top', 'bottom']),
  id: t.maybe(t.String),
  className: t.maybe(t.String),
  style: t.maybe(t.Object),
  menuRenderer: t.maybe(t.Function),
  groupByKey: t.maybe(t.String),
  optionGroupRenderer: t.maybe(t.Function)
};

/** A dropdown field based on [react-select](https://github.com/JedWatson/react-select)
 * @param value - selected value
 * @param valueLink - defines actions to be taken when a particular value is selected
 * @param onChange - called when value is changed
 * @param options - available options
 * @param size - medium | small
 * @param disabled - true if disabled
 * @param searchable - true if it should be possible to search the desired value by writing into the dropdown
 * @param clearable - true if it should be possible to reset the selected value
 * @param backspaceRemoves - whether pressing backspace removes the last item when there is no input value
 * @param multi - true if it should be possible to select multiple values
 * @param flat - whether it should have a flat style
 * @param autoBlur - whether it should blur automatically when the user selects a value
 * @param simpleValue - if true, selected values will be passed to onChange as comma-separated string of values (eg "1,2,3") instead of array of objects
 * @param menuRenderer - the function that can be used to override the default drop-down list of options
 * @param groupByKey - the field name to group by
 * @param optionGroupRenderer - the function that gets used to render the content of an option group
 * @param menuPosition - whether the menu should be rendered on top or bottom when it's open
 */
@skinnable()
@props(Props, { strict: false })
export default class Dropdown extends React.Component {

  static defaultProps = {
    delimiter: ',',
    size: 'medium',
    disabled: false,
    searchable: false,
    clearable: false,
    multi: false,
    flat: false,
    autoBlur: true,
    simpleValue: true,
    groupByKey: 'optionGroup',
    optionGroupRenderer: defaultOptionGroupRenderer,
    menuPosition: 'bottom'
  }

  componentDidMount() {
    this.logWarnings();
  }

  logWarnings = () => {
    if (this.props.children) {
      warn('You\'re passing children. Not expected behaviour');
    }
  };

  getValue = () => (
    this.props.valueLink ? this.props.valueLink.value : this.props.value
  );

  valueToOption = (value, options) => {
    if (t.String.is(value) || t.Number.is(value)) {
      const { multi, delimiter } = this.props;
      if (multi) {
        const values = String(value).split(delimiter);
        return values.map(v => find(options, { value: v }));
      }
      return find(options, { value });
    }
    return value;
  };

  sortOptionsByGroup = (options) => {
    return sortBy(options, (option) => {
      if (option[this.props.groupByKey] === undefined) {
        return '';
      }
      else {
        return option[this.props.groupByKey];
      }
    });
  }

  getOnChange = () => this.props.valueLink ? this.props.valueLink.requestChange : this.props.onChange;

  getCustomClassNames() {
    const { size, flat, clearable, menuPosition } = this.props;
    return cx({
      'is-medium': size === 'medium',
      'is-small': size === 'small',
      'is-flat': flat,
      'is-clearable': clearable,
      'menu-position-top': menuPosition === 'top'
    });
  }

  _onChange = _value => {
    const onChange = this.getOnChange();
    const value = isEmptyArray(_value) ? null : _value;
    return onChange(value);
  }

  onInputKeyDown = (e) => {
    if (e.keyCode === 38 || e.keyCode === 40) {
      if (isEmptyArray(this.props.options)) {
        e.preventDefault();
      }
    }
  }

  focus = () => { this.select.focus(); }

  optionGroupRenderer = ( title ) => {
    if (title === 'undefined') {
      return null;
    }
    else {
      return (
        <FlexView className='Select-option-group'>
          {this.props.optionGroupRenderer(title)}
        </FlexView>
      );
    }
  }

  _menuRenderer = ({
    focusedOption,
    instancePrefix,
    onFocus,
    onSelect,
    optionClassName,
    optionComponent,
    optionRenderer,
    options,
    valueArray,
    valueKey
  }) => {

    const Option = optionComponent;
    const propertyName = this.props.groupByKey;

    return map(groupBy(options, propertyName), (optionGroup, optionGroupTitle) => {
      return (
        <FlexView column key={optionGroupTitle}>
          {this.optionGroupRenderer(optionGroupTitle)}
          {optionGroup.map((option, i) => {
            const isSelected = valueArray && valueArray.indexOf(option) > -1;
            const isFocused = option === focusedOption;
            const optionRef = isFocused ? 'focused' : null;
            const optionClass = cx(optionClassName, {
              'Select-option': true,
              'is-selected': isSelected,
              'is-focused': isFocused,
              'is-disabled': option.disabled
            });
            return (
              <Option
                className={optionClass}
                instancePrefix={instancePrefix}
                isDisabled={option.disabled}
                isFocused={isFocused}
                isSelected={isSelected}
                key={`option-${i}-${option[valueKey]}`}
                onFocus={onFocus}
                onSelect={onSelect}
                option={option}
                optionIndex={i}
                ref={optionRef}
              >
                {optionRenderer(option, i)}
              </Option>
            );
          })}
        </FlexView>
      );
    });
  }

  getLocals() {
    const {
      _onChange,
      onInputKeyDown,
      _menuRenderer,
      props: { className, options, backspaceRemoves, clearable, ...props }
    } = this;

    return {
      ...omit(props, 'valueLink'),
      options: this.sortOptionsByGroup(options),
      clearable,
      backspaceRemoves: t.Nil.is(backspaceRemoves) ? clearable : backspaceRemoves,
      resetValue: null,
      className: cx('dropdown', className, this.getCustomClassNames()),
      value: this.valueToOption(this.getValue(), options),
      onInputKeyDown,
      onChange: _onChange,
      menuRenderer: this.props.menuRenderer || _menuRenderer
    };
  }

  template(locals) {
    return <Select {...locals} ref={select => { this.select = select; }} />;
  }
}
