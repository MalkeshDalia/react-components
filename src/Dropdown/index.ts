import Dropdown from './Dropdown';
export default Dropdown;
export { Props, defaultMenuRenderer } from './Dropdown';
