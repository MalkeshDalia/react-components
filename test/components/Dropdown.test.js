import React from 'react';
import renderer from 'react-test-renderer';

import Dropdown from '../../src/dropdown';

const exampleProps = {
  ...Dropdown.defaultProps,
  id: '12345',
  className: 'fancy-class-name',
  style: { margin: 10, position: 'relative' },
  value: { value: 'test', label: 'Test', foo: 'bar' },
  options: [
    { value: 'test', label: 'Test' },
    { value: 'test1', label: 'Test1' },
    { value: 'test2', label: 'Test2' }
  ]
};

const dropdown = new Dropdown(exampleProps);

describe('Dropdown', () => {

  it('renders correctly', () => {
    const component = renderer.create(
      <Dropdown {...exampleProps} />
    );
    expect(component).toMatchSnapshot();
  });

  describe('getLocals', () => {

    it('computes className', () => {
      const { className } = dropdown.getLocals();
      expect(className).toContain('dropdown');
      expect(className).toContain('fancy-class-name');
    });

    it('doesn\'t alter object value', () => {
      const { value } = dropdown.getLocals();
      expect(typeof value).toBe('object');
      expect(value).toEqual(exampleProps.value);
    });

    it('computes value from a string', () => {
      const dropdown = new Dropdown({
        ...Dropdown.defaultProps,
        value: 'test',
        options: [
          { value: 'test', label: 'Test' },
          { value: 'test1', label: 'Test1' },
          { value: 'test2', label: 'Test2' }
        ]
      });
      const { value, options } = dropdown.getLocals();
      expect(typeof value).toBe('object');
      expect(value).toEqual(options[0]);
    });

    it('computes value from a number', () => {
      const dropdown = new Dropdown({
        ...Dropdown.defaultProps,
        value: 2,
        options: [
          { value: 0, label: 'Test' },
          { value: 1, label: 'Test1' },
          { value: 2, label: 'Test2' }
        ]
      });
      const { value, options } = dropdown.getLocals();
      expect(typeof value).toBe('object');
      expect(value).toEqual(options[2]);
    });
  });

});
