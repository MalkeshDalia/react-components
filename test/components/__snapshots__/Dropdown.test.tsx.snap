// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown renders correctly 1`] = `
<div
  className="Select dropdown fancy-class-name is-medium Select--single has-value"
  style={undefined}
>
  <div
    className="Select-control"
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "margin": 10,
        "position": "relative",
      }
    }
  >
    <span
      className="Select-multi-value-wrapper"
      id="react-select-2--value"
    >
      <div
        className="Select-value"
        style={undefined}
        title={undefined}
      >
        <span
          aria-selected="true"
          className="Select-value-label"
          id="react-select-2--value-item"
          role="option"
        >
          Test
        </span>
      </div>
      <div
        aria-activedescendant="react-select-2--value"
        aria-expanded={false}
        aria-owns="react-select-2--value"
        aria-readonly="false"
        className="Select-input"
        onBlur={[Function]}
        onFocus={[Function]}
        role="combobox"
        style={
          Object {
            "border": 0,
            "display": "inline-block",
            "width": 1,
          }
        }
        tabIndex={0}
      />
    </span>
    <span
      className="Select-arrow-zone"
      onMouseDown={[Function]}
    >
      <span
        className="Select-arrow"
        onMouseDown={[Function]}
      />
    </span>
  </div>
</div>
`;
